ccstools/man/man8/ccs-editpolicy-agent.8
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.36.
.TH CCS-EDITPOLICY-AGENT "8" "May 2009" "ccs-editpolicy-agent 1.6.8" "System Administration Utilities"
.SH NAME
ccs-editpolicy-agent \- Agent for editing TOMOYO Linux's policy
.SH SYNOPSIS
.B ccs-editpolicy-agent
\fIlisten_ip:listen_port\fR
.SH DESCRIPTION
This is an agent program for ccs\-editpolicy ccs\-loadpolicy ccs\-savepolicy ccs\-ccstree .
.TP
listen_ip:listen_port
Listen at specified IP address and port number.
.SH EXAMPLES

# ccs\-editpolicy\-agent 192.168.1.1:10000
.IP
Listen at 192.168.1.1 port 10000 .
.SH NOTES

 Don't run this program unless you know what you are doing.

 This program allows anonymous users to edit policy. No user authentication is performed.

 This program is designed for systems which running ccs-editpolicy is difficult due to resource limitation.
 This program is not designed for PC and servers.

 You need to register either path to this program or a domain for this program in /proc/ccs/manager before invoking this program.
.SH AUTHORS

 penguin-kernel _at_ I-love.SAKURA.ne.jp
.SH COPYRIGHT
Copyright \(co 2005-2009 NTT DATA CORPORATION.
.PP
This program is free software; you may redistribute it under the terms of
the GNU General Public License. This program has absolutely no warranty.
.SH "SEE ALSO"

 ccs-editpolicy (8)
 ccs-loadpolicy (8)
 ccs-savepolicy (8)
 ccs-ccstree (8)


ccstools/ChangeLog
Version 1.0   2005/11/11   First release.

Version 1.0.1 2005/12/08   Minor update release.

  makesyaoranconf.exe:
    Use versionsort() for sorting entries.

  poled.exe:
    Support "search" command.

  poled_old.exe:
    Support "search" command.
    Fix "initializer" checking bug.

  syspol.exe:
    Support editing without resetting cursor position.

Version 1.0.2 2006/02/14   Procedure review.

  savepolicy:
    Support saving "system policy" and "exception policy"
    in addition to "domain policy".

  The following programs were added.

    editpolicy:
      "syspol.exe" "poled.exe" "poled_old.exe" were integrated
      and renamed to "editpolicy".
      This program can edit "system policy", "exception policy"
      and "domain policy".
      Command key assignments were changed.

    checkpolicy:
      A policy validator taken from "poled_old.exe".
      This program was designed for detecting and fixing errors
      in "domain policy".

    loadpolicy:
      A policy reloader.
      This program was designed for loading policy from the disk
      after clearing current policy in the kernel.

    sortpolicy:
      A "domain policy" sorter.
      This program was designed to compress access logs
      generated by "ccs-auditd".
      You can use normal "sort" command for sorting
      "system policy" and "exception policy".

    make_exception.sh:
      A script to create "exception policy".

  The following programs were renamed.

    "remount.exe" was renamed to "remount_rootfs".
    "makesyaoranconf.exe" was renamed to "makesyaoranconf".

  The following programs were removed.

    "poled.exe" "poled_old.exe" "syspol.exe"
    "obsolete_chksymlink" "obsolete_chroot_su"
    "obsolete_lsdir" "obsolete_makelink" "obsolete_movlog"
    "bindtest" "logtest"  "pathnametest" "rofstest"
    "linuxrc_old"

  The following programs for testing TOMOYO Linux's kernel were added.
  They are in the kernel_test directory.

    "sakura_bind_test" "sakura_capability_test"
    "sakura_filesystem_test" "sakura_trace_test"
    "tomoyo_capability_test" "tomoyo_file_test" "tomoyo_info_test"
    "tomoyo_name_test" "tomoyo_port_test" "tomoyo_signal_test"

Version 1.1   2006/04/01   Functionality enhancement release.

  loadpolicy:
    Delete domain for loadpolicy anyway.

  findtemp:
    Now supports for detecting all nonexistent pathnames.

  savepolicy:
    Run twice inside savepolicy itself to include necessary ACLs
    for savepolicy itself.

  The following program for testing TOMOYO Linux's kernel was added.

    "testall.sh"

Version 1.1.1 2006/05/15   Functionality enhancement release.

  The following programs were added.

    ld-watch:
      Monitors /etc/ld.so.cache and updates exception policy.
      This program is used only when updating packages.

    ccs-queryd:
      Monitors /proc/ccs/policy/query for policy violation and
      asks the administrator whether to grant or reject the request.
      This program is used while and after updating packages.

Version 1.1.2 2006/06/02   Functionality enhancement release.

  The following programs were redesigned.

    editpolicy:
      Simplified domain policy handling and removed "save" key.
      All modifications are taken effect immediately.

    loadpolicy:
      Simplified domain policy handling.

    sortpolicy:
      Simplified domain policy handling.

    savepolicy:
      Save all policies by default.

  The following program was removed.

    editpolicy_offline

Version 1.1.3 2006/07/13   Functionality enhancement release.

  The following bugs were fixed.

    editpolicy:
      The "Commands =" line was too wide to show within 80x25 screen.

    checkpolicy:
      Renamed domain for "initializer" was wrong.

Version 1.2   2006/09/03   Functionality enhancement release.

  findtemp:
    Now displays all nonexistent pathnames.

  editpolicy_offline:
    Redesigned to use the same operation manner.
    Saves changes automatically when exiting.

Version 1.3   2006/11/11   First anniversary release.

  The following program was redesigned.

    checkpolicy:
      A policy validator.
      Reads policy from stdin and prints syntax errors with line numbers.

  The following programs were added.

    setprofile:
      Assigns profiles to domains.

    pathmatch:
      Reads pathname patterns and expands them.

    domainmatch:
      fgrep for /proc/ccs/policy/domain_policy .

    ccstree:
      pstree with profile numbers and domain names.

    patternize:
      Reads domain policy and patternize pathnames.

    proxy:
      A tiny TCP port forwarder, binding to local port explicitly
      to allow servers filter based on client's port numbers.

    mailauth:
      An example program for CERBERUS.

    timeauth:
      An example program for CERBERUS, similar to honey.

  The following programs were removed.
  If you need them, please take from version 1.2 .
    "remount_rootfs" "linuxrc"
    "dumplink" "dumpsymlink" "makelink" "makesymlink"

  The following program for testing TOMOYO Linux's kernel was added.

    "tomoyo_rewrite_test"

Version 1.3.1 2006/12/08   Minor update release.

  The following bug was fixed.

    editpolicy:
      PageUp/PageDown keys and screen drawings were not working well
      on some environments due to forcefully setting "TERM=linux".

  The following program for testing TOMOYO Linux's kernel was updated.

    "newns"

Version 1.3.2 2007/02/14   Usability enhancement release.

  Many tools were merged into single source code.
  Policy editor was redesigned.

Version 1.4   2007/04/01   x86_64 support release.

  The following bug was fixed.

    editpolicy:
      Domain flags was wrong if "keep_domain <kernel>" is given.

Version 1.4.1 2007/06/05   Minor update release.

  Single source code was divided into many source code.

  The following bug was fixed.

    checkpolicy:
      "keep_domain" syntax was not checked correctly.

Version 1.4.2 2007/07/13   Bug fix release.

  .init:

ccstools/ccs-editpolicy-agent.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <signal.h>
#include <dirent.h>

static void show_tasklist(FILE *fp, const _Bool show_all)
{
	int status_fd = open(".process_status", O_RDWR);
	DIR *dir = opendir("/proc/");
	if (status_fd == EOF || !dir) {
		if (status_fd != EOF)
			close(status_fd);
		if (dir)
			closedir(dir);
		return;
	}
	fputc(0, fp);
	while (1) {
		FILE *status_fp;
		pid_t ppid = 1;
		char *name = NULL;
		char buffer[1024];
		char test[16];
		unsigned int pid;
		struct dirent *dent = readdir(dir);
		const char *cp;
		if (!dent)
			break;
		cp = dent->d_name;
		if (dent->d_type != DT_DIR || sscanf(cp, "%u", &pid) != 1)
			continue;
		memset(buffer, 0, sizeof(buffer));
		snprintf(buffer, sizeof(buffer) - 1, "/proc/%u/exe", pid);
		if (!show_all && readlink(buffer, test, sizeof(test)) <= 0)
			continue;
		snprintf(buffer, sizeof(buffer) - 1, "/proc/%u/status", pid);
		status_fp = fopen(buffer, "r");
		if (status_fp) {
			while (memset(buffer, 0, sizeof(buffer)),
			       fgets(buffer, sizeof(buffer) - 1, status_fp)) {
				if (!strncmp(buffer, "Name:\t", 6)) {
					char *cp = buffer + 6;
					memmove(buffer, cp, strlen(cp) + 1);
					cp = strchr(buffer, '\n');
					if (cp)
						*cp = '\0';
					name = strdup(buffer);
				}
				if (sscanf(buffer, "PPid: %u", &ppid) == 1)
					break;
			}
			fclose(status_fp);
		}
		fprintf(fp, "PID=%u PPID=%u NAME=", pid, ppid);
		if (name) {
			cp = name;
			while (1) {
				unsigned char c = *cp++;
				if (!c)
					break;
				if (c == '\\') {
					c = *cp++;
					if (c == '\\')
						fprintf(fp, "\\\\");
					else if (c == 'n')
						fprintf(fp, "\\012");
					else
						break;
				} else if (c > ' ' && c <= 126) {
					fputc(c, fp);
				} else {
					fprintf(fp, "\\%c%c%c",
						(c >> 6) + '0',
						((c >> 3) & 7) + '0',
						(c & 7) + '0');
				}
			}
			free(name);
		} else {
			fprintf(fp, "<UNKNOWN>");
		}
		fputc('\n', fp);
		snprintf(buffer, sizeof(buffer) - 1, "%u\n", pid);
		write(status_fd, buffer, strlen(buffer));
		memset(buffer, 0, sizeof(buffer));
		while (1) {
			int len = read(status_fd, buffer, sizeof(buffer));
			if (len <= 0)
				break;
			fwrite(buffer, len, 1, fp);
		}
		fputc('\n', fp);
	}
	fputc(0, fp);
	closedir(dir);
	close(status_fd);
	fflush(fp);
}

static _Bool verbose = 0;

static void do_child(const int client)
{
	int i;
	int fd = EOF;
	char buffer[1024];
	/* Read filename. */
	for (i = 0; i < sizeof(buffer) - 1; i++) {
		if (read(client, buffer + i, 1) != 1)
			goto out;
		if (!buffer[i]) {
			char *cp = strrchr(buffer, '/');
			const _Bool ps = !strcmp(buffer,
						 "proc:process_status");
			const _Bool ps_all = !strcmp(buffer,
						     "proc:all_process_status");
			if (ps || ps_all) {
				FILE *fp = fdopen(client, "w");
				/* Open /proc/\$/ for reading. */
				if (fp) {
					show_tasklist(fp, ps_all);
					fclose(fp);
				}
				break;
			}
			if (!cp)
				cp = buffer;
			else
				cp++;
			/* Open for read/write. */
			fd = open(cp, O_RDWR);
			break;
		}
	}
	if (fd == EOF)
		goto out;
	/* Return \0 to indicate success. */
	if (write(client, "", 1) != 1)
		goto out;
	if (verbose) {
		write(2, "opened ", 7);
		write(2, buffer, strlen(buffer));
		write(2, "\n", 1);
	}
	while (1) {
		char c;
		/* Read a byte. */
		if (read(client, &c, 1) != 1)
			goto out;
		if (c) {
			/* Write that byte. */
			if (write(fd, &c, 1) != 1)
				goto out;
			if (verbose)
				write(1, &c, 1);
			continue;
		}
		/* Read until EOF. */
		while (1) {
			int len = read(fd, buffer, sizeof(buffer));
			if (len == 0)
				break;
			/* Don't send \0 because it is EOF marker. */
			if (len < 0 || memchr(buffer, '\0', len) ||
			    write(client, buffer, len) != len)
				goto out;
		}
		/* Return \0 to indicate EOF. */
		if (write(client, "", 1) != 1)
			goto out;
	}
 out:
	if (verbose)
		write(2, "disconnected\n", 13);
	close(fd);
	close(client);
}

int main(int argc, char *argv[])
{
	const int listener = socket(AF_INET, SOCK_STREAM, 0);
	struct sockaddr_in addr;
	socklen_t size = sizeof(addr);
	char *port;
	if (chdir("/proc/ccs/") && chdir("/sys/kernel/security/tomoyo/"))
		return 1;
	{
		int i;
		for (i = 1; i < argc; i++) {
			if (strcmp(argv[i], "--verbose"))
				continue;
			verbose = 1;
			argc--;
			for (; i < argc; i++)
				argv[i] = argv[i + 1];
			break;
		}
	}
	if (argc != 2) {
usage:
		fprintf(stderr, "%s listen_address:listen_port\n", argv[0]);
		return 1;
	}
	port = strchr(argv[1], ':');
	if (!port)
		goto usage;
	*port++ = '\0';
	memset(&addr, 0, sizeof(addr));
	addr.sin_family = AF_INET;
	addr.sin_addr.s_addr = inet_addr(argv[1]);
	addr.sin_port = htons(atoi(port));
	if (bind(listener, (struct sockaddr *) &addr, sizeof(addr)) ||
	    listen(listener, 5) ||
	    getsockname(listener, (struct sockaddr *) &addr, &size)) {
		close(listener);
		return 1;
	}
	{
		const unsigned int ip = ntohl(addr.sin_addr.s_addr);
		printf("Listening at %u.%u.%u.%u:%u\n",
		       (unsigned char) (ip >> 24), (unsigned char) (ip >> 16),
		       (unsigned char) (ip >> 8), (unsigned char) ip,
		       ntohs(addr.sin_port));
		fflush(stdout);
	}
	close(0);
	if (!verbose) {
		close(1);
		close(2);
	}
	signal(SIGCHLD, SIG_IGN);
	while (1) {
		socklen_t size = sizeof(addr);
		const int client = accept(listener, (struct sockaddr *) &addr,
					  &size);
		if (client == EOF) {
			if (verbose)
				fprintf(stderr, "accept() failed\n");
			continue;
		}
		switch (fork()) {
		case 0:
			close(listener);
			do_child(client);
			_exit(0);
		case -1:
			if (verbose)
				fprintf(stderr, "fork() failed\n");
			close(client);
			break;
		default:
			close(client);
		}
	}
	close(listener);
	return 1;
}
